Haml coffee

vendethiel opened the pull request:
not sure what happened but it was a cool discussion. (discussion via github is a bit painful over time but that's romantic enough for today, y'see)
I'm back, no more with, no more escape (as much as possible).
With my version : 

```
render: 147.000ms
initial post insertion handlers: 79.000ms
interactive: 325.000ms 
```

with your version : 

```
initial render: 138.000ms script.js:1268
initial post insertion handlers: 83.000ms script.js:1289
interactive: 305.000ms 
```

(I got sometimes result worse for your version, sometimes 50ms better, but for 10 tries I got around 320-330ms for the first and 295-310 for yours).

vendethiel commented:
btw it should have rebased itself somehow haha not sure maybe that's just still as broken and i removed your diffs idk.
I tried it on the large /q/ board. Still too many `attrs()` calls I think :(.

qqueue commented:
This isn't based on my master, so you don't have my slightly improved metrics, among other things.

Rebase and fix the conflicts, then force-push your branch.

#4 is still there, can't you just push to that PR?

vendethiel commented:
> #4 is still there, can't you just push to that PR?

Nope, maybe because you closed me

> Rebase and fix the conflicts.

I did my best, but apparently I just failed haha :D I'll let you call the family to announce the news
if you can fix this effin' mess, I'd looooove to hear that, cause git is cool when it works but else i have NO IDEA WHATSOEVER what went wrong. And squashing is a no-go because windows

qqueue commented:
Whatever you did, the conflicts are gone.

I'll test this in a bit, doing "real" work right now. If you want to run the comparisons yourself and paste the stats here, that would help.

If I merge this, I'm going to squash anyway, but in the future, please improve your commit messages:

http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

Not mentioned in the article, but still important: use imperative tense, e.g. "Fix some shit", not "Fixed some shit". See my commit messages after August 2012 or so for good examples, before that for bad.

vendethiel commented:
> Whatever you did, the conflicts are gone.

This part is ok, but I was not sure you'd be able to merge cleanly anyway

>  but in the future, please improve your commit messages:

Windows. I can't split my commit messages on several lines, else I'd. Oh, right, I can use the -F option to pass a file but seriously, that's a pain. I do that sometimes, when I think it's important but not when it's for a PR that's gonna get squashed. If you actually want me to do so, fine, just tell me beforehand

> Not mentioned in the article, but still important: use imperative tense, e.g. "Fix some shit", not "Fixed some shit"

Isn't that what I did ? "convert", not "converted". Also I wrote "as using", not "at using" :(.

qqueue commented:
> Windows. I can't split my commit messages on several lines

![horrifying](https://f.cloud.github.com/assets/1281194/158867/3979d010-76f8-11e2-82e5-5d14b72231e6.png)

Either use Github's thing or tortisegit, or fix your $EDITOR so 'git commit' will launch one for you.

---

> imperative tense

"Convert" is better than "converted", but overall I'd prefer the commits to be more like:

```
Rewrite board.template.html as jade
Rewrite other templates as jade
Remove jade inefficiencies for faster rendering
Fix typo in template
```

or better yet, one commit:

```
Use jade templates instead of string interpolation 

squashed commits:

...
```

Then I could merge from github.

vendethiel commented:
> Either use Github's thing or tortisegit

That's a no-no, both are crappy

> or fix your $EDITOR so 'git commit' will launch one for you.

Despite all my googling, I never managed to make that run on windows/sublime text. Always has starting issues, path issues, "no file COMMITMESSAGE" issues ETC.

> I'd prefer the commits to be more like:

I'd say you're the kind of guy with colorful comments and I'm one with colorful commit messages :p. Must admit I don't do that offen, not really willing to spend 20min in my git diffs to check for these, i have my changelog for that :p. 
But I'll remember that ...next time

> or better yet, one commit:

Would be cool, right? Oh, but, windows. Even if it meant being your valentine, i'd refuse to squash on windws.

qqueue commented:
![1359264778509](https://f.cloud.github.com/assets/1281194/158945/7bd98ede-76fb-11e2-82b1-e558a0f3d73a.jpg)

Sounds like you need to get a **_REAL MAN'S OPERATING SYSTEM**_, son.

vendethiel commented:
Sounds more like I need to get my own pc.

qqueue commented:
On the big /q/ sticky thread, 897 posts, using my netbook:

```
init                            : 186ms
parse page                      : 10ms
parse thread                    : 963ms
generate new HTML body          : 110ms
parse and render new body       : 4514ms
set new page title              : 2ms
set correct favicon             : 20ms
initial render                  : 5681ms
initial post insertion handlers : 12837ms
interactive                     : 25382ms
```

With Jade:

```
init                            : 191ms
parse page                      : 12ms
parse thread                    : 809ms
generate new HTML body          : 401ms
parse and render new body       : 4499ms
set new page title              : 2ms
set correct favicon             : 17ms
initial render                  : 5799ms
initial post insertion handlers : 13101ms
interactive                     : 26194ms
```

`attrs` still takes up a majority of the time:

```
Calls    6537
Percent  53.77%
Own Time 351.312ms
Time     389.735ms
Avg      0.06ms
Min      0.032ms
Max      1.083ms
```

Thoughts:
- Jade is still about 3x slower :(
- My main bottleneck to interactive is actually the "forcequotes" feature,
  which does pretty heavy DOM manipulation to compact quote links and add
  forced quotes. I knew that already though; still need to think of a better
  way to do forced quotes.
- The second bottleneck to interactive is likely just pure firefox parse/render
  time. The actual generation of an HTML string is only ~5% of the 'initial
  render', and a smaller fraction of 'interactive' (but that includes
  forcequotes).
- I only have three templates now, but as html5chan grows, I can see a need for
  more (mostly revolving around cool d3 graphs). When I do have more, I'd
  prefer to keep them in the same language, be it string interpolation or jade.
- The two parts of manual string interpolation that I find most annoying are
  handling `undefined` as `''` and dealing with larger conditional block
  indentation. The HTML part of templating is bearable. If coco had better
  handling of undefineds (maybe a `#[that]` syntax that desugars to `#{if that?
  then that else ''}`, like `#{[that]}` but without the perf hit), I'd be more
  okay with string interpolation.
- But similarly, if jade was only 50% slower, i.e. if `attr` could be
  eliminated, I'd be okay with jade.

I still don't really merge now and add overhead when I'm not actively messing
around with HTML anyway, but I'll leave the issue open. If you can get rid
of `attrs`, I think it will be worth it overall.

vendethiel commented:
6k5 calls ? The fuck.
Look at that [gorgeous html that haml-coffee generates](https://github.com/Nami-Doc/wowboardhelpers/blob/master/wowboardhelpers.user.js#L131). I don't have any benchmarks but that should be the hell faster than jade
If you prefer that (and you're ok with with haml), I can copypaste from my slakefile+`s/(\n\s+)([a-z])/$1%$2/` the templates (and I'll run benchmarks to compare w/ my original PR).

EDIT : no i'm not stressing to do a pr blablabla

qqueue commented:
Is haml-coffee's syntax much worse than jade?

Aside: do you remember the jQuery bug in which you found [ `<span class=""></span>` behavior](https://github.com/qqueue/html5chan/pull/4#issuecomment-13517762) being discussed? I tried googling for it, but I couldn't find anything and I'm still curious.

vendethiel commented:
Didn't even tried. Somebody asked why he wasn't getting elems such as `<span></span>` with `[class='']` or something like that.

hade | haml :

``` jade
head
  title Hey
body
  h1
    #container.
      How are you ?
```

``` haml
%head
  %title Hey
%body
  %h1
    #container
      How are you ?
```

tags are prefixed with %, everything else is text

One thing about haml-coffee tho,I don't think you can use a dynamic tag

vendethiel commented:
want me to hamlcoffeeize? or maybe hamlcocoize or watevar. I don't think we can use dynamic tags for now, but that can probably change.

qqueue commented:
Sure, if you think it's performant enough. Gotta go fast.

Re: dynamic tags, the only place I'm using it is to switch thread OP's to `<div>`'s instead of `<articles>`. Since the thread is an `<article>` on its own, the thread needs to inherit the OP's `<h1`> directly to follow the HTML5 spec (all sectioning elements must have titles). That's just HTML5 outlining pedantry though, so it doesn't really matter.

vendethiel commented:
better nerf string interp. lemme finish my spice&wolf and i'll be all yours.

and you reminded me I should use something actually correct for stylus :/.

vendethiel committed:
merge conflicts

vendethiel commented:
Fuck you git, seriously, fuck you, piece of crap

vendethiel commented:
please tell me it's okay, that my pain is to end and that we'll live happily ever after

vendethiel committed:
remove conflict artifact

vendethiel commented on Cokefile:
```diff
@@ -19,12 +21,37 @@ task \build 'build userscript' ->
+        escape-html: false
+        escape-attributes: false
+        uglify: true
+#            customCleanValue: 'c$'
```
I can add it back and do that in coco. I love `Function::toString` :D.

``` coffee
               "" + ->
                  if document.getElementById \captcha
                    Recaptcha
                      &_init_options do
                        theme: \custom
                        custon_theme_widget: that
                      &theme = \custom
                      &widget = that
                      &_finish_widget!
```

qqueue commented:
Unscientific test with netbook on http://boards.4chan.org/a/res/81339111 (at ~204 posts/42 images):
- string interp `generate new HTML body: 40ms`
- hamlc `generate new HTML body: 237ms`

Not exactly sanic speed :snail:

vendethiel commented:
meh I can't even browse topics, you're trying to access to ".boardTitle" which does not exist for me

very interesting speed benchs :)
![abc](http://i.imgur.com/GjOQLqI.png)

my script metrics : 

```
generate new HTML body: 16.000ms script.js:1847
initial render: 66.000ms script.js:1862
interactive: 149.000ms 
```

yours

```
generate new HTML body: 3.000ms script.js:1788
initial render: 53.000ms script.js:1803
Uncaught ReferenceError: get is not defined script.js:2185
Uncaught TypeError: Cannot read property 'length' of undefined script.js:1665
interactive: 121.000ms 
```

and http://boards.4chan.org/a/res/81339111 just fails because `$ \.boardTitle` is null.

vendethiel commented:
hmm actually I have another idea now.

``` coffee
jade = require \jade
compile-jade = (it, filename) ->
  jade.compile it, {filename}
# ...
\.cojade : (src, filename) ->
  src.=toString!
  src.=replace /#{/ /{{/
  """
module.exports = \"""
#{compile-jade(src)!replace /{{/g /\#{/}
\"""
  """
```

So that
`html #{abc}`
becomes
`html {{abc}`
which jade will compile
`<html>{{abc}</html>`
then replacing again
`<html>#{abc}</html>`
and the final step "cojade step"

``` coffee
module.exports = """
<html>#{abc}</html>
"""
```

Could maybe use `+pretty` too :).
The idea is very appealing because the perf hit should be inexistant (out of compiling ofc). A drawback would be that we can't use the simple form anymore : `html= abc`.

Or maybe we can. `.replace /([a-z\)\}])=\s+(.*)$/g "$1 #{$2}"` before converting `{{` and passing it to jade.

Does that sound interesting to you too ? Would you rather see jade or haml then ? I'd probably say jade for myself, so that would be it for the PR too.
I was actually fine with the numbers I had on my own script (considering I have at most 20 posts or 20 threads) but this idea seems interesting.

I just need to figure out how I can make it work for the dynamic stuff (if, for, etcie) eheh...

vendethiel commented:
Spent around 1h doing really bad stuff : https://gist.github.com/Nami-Doc/8595f570f2bd294e5285

vendethiel commented:
just realized this wasn'"t working at all when dedenting more than once ... fuck it :)

qqueue commented:
> `$ \.boardTitle` is null

Hmm, moot must've been messing around. I'll have to remember to guard access to it.

---

Maybe we should just design our own templating engine for our very specific "HTML with sanic speed, no escaping" use case. Even very light compilation to get rid of the boilerplate in my current `board.cohtml` would be a good start. Then we could develop fancier indent-based compilation later. It'd be also cool to emulate some of emmet/zencoding style, e.g.:

``` html
for posts
    <h1>{&title}</h1>
    <div id="c{&no}" class="post {if &preview then 'preview'}">
        {&comment}
    </div>
```

->

```
{posts}*(h1{title} + div#{c{&no}}[class="post {if &preview then 'preview'}"] > {&comment})
```

vendethiel commented:
I find that to be reading insanely bad. If I lose perfs AND lose readability ... ye no. Plus adding "nullcheck" in my case is rather easy : just add `(` before the `#{` and `) ? ''` after the `"""`.

And I managed to make it work, it seems.
https://gist.github.com/Nami-Doc/8595f570f2bd294e5285
EDIT : actually, lemme auto-add the join ... hehe
EDIT2 : working well enough. Will convert my templates later

may I ask you what time is it ?

qqueue commented:
>not liking zencoding
But it's HTML as perl, man. Just look at that alpha:symbol ratio!

vendethiel commented:
Actually, I'm quite ok with zencoding... I just find jade to be far better.
was traumatized enough by perl for this not to be a compelling argument. The hours I spent trying to debug obscure things in my ircbot when I was 12 :(.

vendethiel commented:
Just shaved off ~15ms of my 80ms using that. And I have all that sugar goodness from co{ffee,co}/ls

qqueue commented:
Er,  you keep editing your posts, so it's hard to tell what's current. In any case, I broke your changes again, sorry. I pulled out my current templates to separate `.cohtml` files though, so it should be easier for you to make a clean patch if you want.

vendethiel commented:
> Er, you keep editing your posts, so it's hard to tell what's current.

Yeah I don't want to spam you 

> In any case, I broke your changes again, sorry. I pulled out my current templates to separate .cohtml files though, so it should be easier for you to make a clean patch if you want.

I saw that, but I still think you're a meanie and you're doing that on purpose. You know by now how much I _hate_ it
I might do it but that won't be for now :(. ho well still 1 week holidays~.

What do you think of my "work" anyway (do tell it's clever)?

qqueue commented:
> I don't want to spam you

I'd prefer it, actually. It keeps the message timeline linear and easy to follow.

> you're doing that on purpose

Alpha-quality software, dude. Having the current templates in separate files should make your patch easier now though, since you only need to change them and the Cokefile. But no promises.

> What do I think

Definitely clever. If I were as motivated as you to bring jade templates to html5chan, I'd probably make the same hacks. In the longer term, I'd rather have a "real" template compiler though, because it'd more maintainable than regex hacks.

vendethiel commented:
> I'd prefer it, actually. It keeps the message timeline linear and easy to follow.

Yeah. It might hurt your overall timeline tho, but your own is probably far better than mine haha

> Alpha-quality software, dude

I actually heard that with the "but duuuuude" voice from yugioh abridged.

> as motivated as you to bring jade templates to html5chan

That's not really what motivates me. I just found : "whoah, I need to try to do that - it seems like a problem interesting enough to solve".
And I'm not using any regexp hack. I already have too much of those with your "linkifier" haha...

> In the longer term, I'd rather have a "real" template compiler though

Obviously. But I'm far too lazy.

And you know what ? Jade accepts a `compiler`. I can probably override the [`visits`](https://github.com/visionmedia/jade/blob/master/jade.js#L237). Stills seems a bit complicated BUT far better in the long term run.

vendethiel commented:
Well, I just fixed the last problem. (not that I [tried too hard to fix it](https://github.com/Nami-Doc/wowboardhelpers/commit/98a331f96d46268a16da6ce01e3a68349b127957#L0R53)).
It's currently pretty ugly. I could probably stop incrementing the indent all the time and use `<=` in the while. Actually I think the soaked if is useless uh. the `prev-indent` could probably be removed for a smarter `indent - extra-level - 1` or something like that...
I don't want to publish that as a npm module, that's disgusting. I don't mind because I had fun debugging it.

My questions are
1) Would you prefer it in coco ? (that should minor to change tho)

2) Are you okay with how it's gonna look look ? Luckily enough, closing the other PR set the commits to stone, so I [still have the jade files](https://github.com/Nami-Doc/html5chan/tree/d743adaef1f31948953b0842a463924df7522a77/templates). They should work fine EXCEPT that attributes should be `(b="#{c}")` and not `(b=c)` (to avoid jade requiring it runtime).
And it's not going to be `each thread in threads` but `for thread of @threads` (and things like that).

3) Are you okay with how the code looks ?
What I'm doing inside of "jadels-to-ls" is basically :
- Prepare it : 
  - `@` => `locals.`
  - `#{` => `{{`
- Convert it : 
  - Fix indent
  - Use jade's `|` text for "tags" (flow structures)
  - Null check interpolations (`or ''`)
- Normalize it : 
  - `{{` => `#{`
- Compile LS
- Wrap in into a `module.exports`. (and in my case, a fun

4) I don't remember right now what I wanted to say :D.

qqueue commented:
1. Yes, because then I only need to have one transpiler in `package.json`
2. jade syntax is definitely better than what I have currently (`.cohtml`). My only real concern remains runtime speed.
3. Wary of the fragility, but I have ~80 lines of templates. Not a huge deal if I need to switch things up.
4. Will it handle partials?

vendethiel commented:
> jade syntax is definitely better than what I have currently (.cohtml). My only real concern remains runtime speed.

Have you looked at the JS generated ? Are you okay with it ;) ? (that should be the SAME as the one you'd write).

---

And in your case, it might be bothersome to have to wrap attributes in `"#{}"` (ie `a(href="#{@href}")`). I'd like to avoid MORE mess but maybe I'll resort to creating a function which will ensure we have something right.

``` ls
if line ~= /([\t\s]*[a-z{}])\(/
  attrs = line.slice that.0.length
  attrs.slice 0 attrs.indexOf ')' # that's stupid.
  for attr in attrs / ','
    attr.=trim!
    if attr.0 not in <[" ']>
     #enquote it
```

Or maybe I could match, inside `()` (only finding the next `)` 'cause laziness); try to see if `=` is followed by `<[ " ' ]>` and if it's not, add `"{{` (and `}"` before the next `,`).
brrrr. The main problems (cause the code is crap already) is that this fails to handle `abc()` as an attr. But I think jade fails pretty bad already, ie `a(b="c)")` fails)
Or maybe I could rewrite `attrs()` calls. Ha !

---

> Will it handle partials?

Just pass in the template function?

---

And currently, it only recognizes tabs because I use tab. That will need a bit more tweaking, probably a line or two...

vendethiel commented:
fixed another bug, that "indent debt" is really pissing me off.
Got the board template to compile, but I added the `!{`&`}!` style interp (removed by now ... I prefer `a(b="#{c}")` rather than `a(b=!{c}!)`. But I added tag interpolation (`#{container}`). It adds a bit of complexity for a result I'm not sure to approve but well...
I am a bit reluctant to try to find `(` just after `[a-z]+` (partly because of tag interp, even if it should not add much complexity) because I need to find the right closing `)` and that may not be the one I think. (then I could split on `,`, trim, and add `"#{` and `}"`)

I never knew doing wrong things could be so much fun. Maybe I'll try to add a goto ;).

vendethiel committed:
cojadify

vendethiel committed:
.cojade files

vendethiel commented on templates/post.cojade:
```diff
@@ -0,0 +1,44 @@
+  mod="#{if @capcode and @capcode isnt '## Admin' then ' mod' else ''}",
+  uid="#{if @uid then 'uid' else ''}"
+  data-tripcode="#{if @tripcoded then that else ''}",
+  data-idx="#{@idx}", id="#{id or 'p' + @no}")
```
this is horrifying. Let's find a way to kill it with fier§§

vendethiel commented on templates/post.cojade:
```diff
@@ -0,0 +1,44 @@
+    button.hide(type="button", value="#{@no}") &times;
+    button.report(type="submit", form="reportform",
+      class="report", name="no", value="#{@no}") !
+    a.subject(href="#{@url}"))= @subject
```
just in  case you missed it, 'tis auto-`or ""`ed

vendethiel commented:
So now you have a better version than me. Ha!
You can't use `"` unescaped in attributes. That seems normal (tho not coco-alike).
All text `= abc` is auto `or ""`ed like I said in the comment. The for is auto-joined.
It compiles, even tho I havn't tested it yet.
Here is what compiled post.cojade looks like & a compiled version : 

https://gist.github.com/Nami-Doc/4118b21666d5602eca05

EDIT : Which made me realize I used "isnt". Fixed.
EDIT2 : adding the precompiled version made me realize `join` should account for null. Fixed.

qqueue commented:
> error: id is not defined

your templates only compile with one argument, but the post template takes several.

vendethiel commented:
It should be `@id` since I'm mixing these. Not sure what's the best here.... Pass a second `extra` array ? (`@@`).
And I'm starting to think a safe interp would be cool. (maybe `#[]` ? but that'd mean regexp hack probably)
I could probably `.replace /=@([a-zA-Z\._-]+)(,|\))/ '="#{$1 or ''}"$2'` too. That's a regexp hack :(. But that'd clean it up a lot I think.

I also need to add `+debug` to jade

vendethiel commented on templates/board.cojade:
```diff
@@ -0,0 +1,69 @@
+    if board.page
+      li: a(href="#{board.page - 1}") previous
+    li: a(href="#{board.url}") 0
+    li: a(href="1") 1
```
... why am I not using jade for loop here ? I'll add `--` to convert to jade's `-` (since our `-` is co interpolation, that's how partials are done). Not sure about the sigil. Maybe something like `%` ? Or use that as our interpolation ? You tell me.
Aaand that should be `@board` I think

vendethiel commented:
I tried to run the test here but
- On chrome I get an error for board since you're not allowed to `export` stuff
- On firefox I get 

```
[13:13:59.243] [html5chan@httpsgithub.comqqueuehtml5chan] Component returned failure code: 0x80004003 
(NS_ERROR_INVALID_POINTER) [nsIDOMHTMLDocument.replaceChild] @ chrome://scriptish/content/scriptish.js -> 
file:///C:/Users/User/AppData/Roaming/Mozilla/Firefox/Profiles/sc0sqzsi.default/scriptish_scripts/html5chan@httpsgithubcomqqueuehtml5chan/html5chan@httpsgithubcomqqueuehtml5chan.user.js:1710
```

That line being 

``` javascript
        d = document.replaceChild(html, document.documentElement);
```

Maybe that's because of scriptish ?

Switched to firefox, I have an error `board.locals is undefined` but the line isn't even in the "show soruce"...

vendethiel committed:
fix @, coco interp is now `~,`, rendering works

vendethiel committed:
correct pagination

vendethiel commented:
Aaaand my metrics : 

/a/

```
generate new HTML body: 9ms
parse new body HTML: 80ms
prerender handlers: 0ms
render new body: 2ms
highlight current page: 7ms
set new page title: 0ms
set correct favicon: 2ms
initial render: 276ms
onready handlers: 6ms
interactive: 470ms
```

Tried the /q/ sticky thread and the one you linked earlier but they're archived on foolz :(

qqueue commented:
Cool, will test this in a bit.

> On chrome I get an error for board since you're not allowed to export stuff

Hmm, that's probably because chrome's `window` isn't global `this`. I'll have to change the `window.board =` in html5chan.co.

> Maybe that's because of scriptish

Scriptish is a dead fork, you know. You should be using GM 1.3+.

---

Total aside: I realized I can track some html5chan usage through the youtube API (which requires a registered API key):

![Screenshot - 03182013 - 05:11:43 PM](https://f.cloud.github.com/assets/1281194/273391/babe9b3c-9021-11e2-9d54-8ef82f367d9a.png)

The big spike is when I had a bug in the youtube loader that hit the API with empty requests over an over again, whoops.

That FR is probably all you--not sure who the DE and GB are.

qqueue commented:
It runs, but the output isn't correct:

![Screenshot - 03182013 - 08:02:41 PM](https://f.cloud.github.com/assets/1281194/273832/30a1ffc6-9039-11e2-8fdb-89cfa18a3b4b.png)

Am I missing something? Ignore the horrible font rendering, I'm working on it.

vendethiel commented:
Thread.cojade L12, is it the wrong container ?

vendethiel commented:
> Hmm, that's probably because chrome's window isn't global this. I'll have to change the window.board = in html5chan.co.

Yep, I had to migrate all my code to `require()`s for it to work again

> Scriptish is a dead fork, you know. You should be using GM 1.3+.

Changed by now (didn't worked under scriptish). But GM sucks hard

> That FR is probably all you--not sure who the DE and GB are.

Élémentaire, mon cher Watson.

qqueue commented on templates/post.cojade:
```diff
@@ -0,0 +1,44 @@
+#{@container}(data-no="#{@no}", class="#{@classes}", post="post",
+  imagepost="#{if @image then 'imagepost ' else ''}", sage="#{if @sage then 'sage ' else ''}",
```
These should be classes, not HTML attributes.

vendethiel replied:
Whoops. Oversight on my part

vendethiel replied:
hummm... I will move somewhere...

qqueue commented on templates/thread.cojade:
```diff
@@ -0,0 +1,12 @@
+~ post = require \./post
+
+#{@container}(id="#{@id or "t#{@no}"}', data-no="#{@no}",
+  class="#{@classes} thread #{if @sticky then 'sticky ' else ''} #{if @closed then 'closed ' else ''}")
```
@classes can be undefined.

qqueue commented on templates/thread.cojade:
```diff
@@ -0,0 +1,12 @@
+      a.expand-link(href="#{@url}") Expand
+  .replies
+    for reply of @replies
+      = post {container: \article classes: \reply ...reply}
```
This should be correct. I am a bit concerned about the performance of the object mixin `...`, but I'll wait for stats.

vendethiel replied:
Yep, that's what I said, and this is why I proposed `@@` as `extra.` (2nd arg)

vendethiel committed:
soak classes, classes <- data attributes, remove object mixin

vendethiel commented:
I'm still unsure about that `gen-classes.co`, but leaving that in the template was a no-go.

qqueue commented on templates/thread.cojade:
```diff
@@ -0,0 +1,12 @@
+~ post = require \./post
+
+#{@@container}(id="#{@id or "t#{@no}"}', data-no="#{@no}",
+  class="#{@classes ? ''} thread #{if @sticky then 'sticky ' else ''} #{if @closed then 'closed ' else ''}")
```
Not sure what's wrong here, but the thread always comes out with id or class (plain `<article>`)

vendethiel replied:
I'll probably move that into the new helper file

vendethiel replied:
EDIT2 : actually, I know why, I'm just being stupid. I'm not allowed to use `"` in the `id=` interp.

qqueue commented on templates/post.cojade:
```diff
@@ -0,0 +1,41 @@
+    button.hide(type="button", value="#{@no}") &times;
+    button.report(type="submit", form="reportform",
+      name="no", value="#{@no}") !
+    a.subject(href="#{@url}"))= @subject
```
I think there's a superfluous `)` here, because every post has ")" as a subject.

vendethiel committed:
most topic classes, fix interpolation, fix extra paren

vendethiel committed:
remove extra debug

vendethiel commented:
Works fine with this commit :p

qqueue commented:
Fixed some other issues, now it works fine. I don't have hard statistics, but haml-coco is consistently about  ~10% slower in HTML generation time, which is negligible in initial render time fortunately.

Thanks for the hard work!

vendethiel commented:
That's jade coco, but yay~ :p.

> but haml-coco is consistently about ~10% slower in HTML generation time, which is negligible in initial render time fortunately.

Even if you said 10% was an okay threshold, I'd be interested to know where it's slower ? Auto-soaking everywhere ?

qqueue commented:
Some guesses as to why it's slower:
- `variable || ''` rather than 'variable ? variable : ''`
- extra arguments passed as `extras` object literal (that has to be allocated), instead of simple positional arguments
- `require` call in every templated post, rather than as a closed-over variable

Too lazy to test it through. I'll leave that up to you.

vendethiel commented:
> variable || '' rather than 'variable ? variable : ''`

Uh. That's totally negligible... EDIT : http://jsperf.com/or-vs-ternary-cond

> extra arguments passed as extras object literal (that has to be allocated), instead of simple positional arguments

Meh, 10% for that? Hardly believable, v8 optimizes that creating classes (better than anything else since it's not polymorphic)

> require call in every templated post, rather than as a closed-over variable

Good point, that's probably it. Just pass it in extra then :-).
But I find the result pretty nice =).

qqueue commented:
I changed `@@` to positional arguments, and factored out `require` calls inside templates. I did both in one commit so I'm not sure of the specific contribution of each change, but the end result is now just as fast as my old string interpolation.

Passing in the template functions is annoying though. Need to think about how to manage that better.

vendethiel commented:
I see your actions the second you do them. I'm doing a little bench right now.

vendethiel commented:
You could probably move the first liners `~` statement outside the closure returned, for example ;)

qqueue commented:
You should stick jade-coco in its own repo/npm module. Then I can open issues on _your_ projects.

vendethiel commented:
[How could I not think about it before.](https://github.com/Nami-Doc/cojade-to-jade)

It's not updated with the current version but I'm clearly not liking your argument inlining :(. I know it's gotten to a point where the code is so bad anyway more mess can't be that bad but...
(I agree a solution is needed)
But if you're happy with what's I've done, it's nice to start with :sparkling_heart:

vendethiel commented:
People's live will improve https://github.com/ForbesLindesay/with
